### ISSUE TITLE
Can you add controller velocity to IVRData?

### ISSUE THREAD

**Issue by yohlimem** [hammy3502/mc-vr-api#27]
I think that this will make it much easier to code with the API.

**Comment by hammy275** [comment]
I've looked into it before, and it's honestly a good idea. The big issues I would want to hear feedback on (and am also thinking on myself) would be:

- How much past controller information do we store? To calculate velocity, we need to store information from previous ticks, and different mods may want different accuracies on that velocity. The solution for this I've considered so far is each mod declares at some point how many ticks it wants to go back, and the highest number reported is what we keep. However this seems pretty ripe for us getting large numbers that mods definitely don't need, but if we put a cap on the maximum, we might as well just use that as the amount we store anyways.
- What math would we use for calculating velocity for data beyond two ticks? There may be an obvious answer here, since I haven't done any research into the area, but I don't know what math this would entail.

On the whole though, this would be _extremely_ useful. I already do things like this in ImmersiveMC for several immersives, and I think Viveecraft keeps a history of past tick information for similar purposes.

I wouldn't have too much issue adding this to all `IVRData` instances, since the majority of them are client-only, and we shouldn't be taking much memory just storing a bunch of `double`s. I'm wondering though how many of them people would even use velocity for (pre-tick, post-tick, pre-render, post-render, etc.)?

**Comment by yohlimem** [comment]
To calculate the average velocity, you need to subtract the starting and ending points of the IVRData and divide by the time that has passed (chat GPT told me hehe). This will require storing an array of positions that have been measured since the start tick of the world, with an upper limit of 328 positions (8KB) being sufficient. It's worth noting that there's generally no need to calculate velocity over a long duration, so there's little benefit to checking velocities from 18 seconds ago or longer (328 / 20). Similarly, checking velocities over a very short time frame, like two seconds, may not be used that much since one second is at most a very slow throw or swing.

**Comment by hammy275** [comment]
Please don't provide answers from ChatGPT. There's a lot mixed in there that makes literally no sense for the context at hand.

In response to the main point of how it recommends calculating velocity:
"Subtract[ing] the starting and ending points" isn't really great here imo, as beyond _very_ short timescales, this wouldn't really work. Think if the controller is moving in an arc on a 2D grid (for sake of example, this of course also works in 3D), if the controller is at the same height at the two times we use to measure velocity, we're going to get a completely different direction for the velocity in comparison to where we're actually going. Here's a very crude paint sketch of what I mean ![image](https://user-images.githubusercontent.com/29229128/223908157-023a9b5e-d7c8-46ed-b3c3-2d8a31a1d586.png)
EDIT: Admittedly, the "Actual velocity" here is really the "Most likely actual velocity", as there's technically nothing stopping a human hand from suddenly pivoting in a different direction. However, I think for a velocity API such as this, we can pretty safely assume that people will make natural movements.

Problematically, throws can happen very quickly or very slowly. When adding throwing to ImmersiveMC for example, I found the sweet spot for quick throws to be to use the position from two ticks ago to the current tick. I got very weird results when using results from even 4 or 5 ticks ago and funny inaccuracies from comparing to simply last tick (too short of a distance, so hands not moving in perfect motions causes funny directions), leading me to believe that if we want a truly rugged velocity API here, we need to be doing something far more complex than average velocity between two points. I'll be doing more research into this to figure out what's ideal here, but average velocity is not really the ideal solution I'd want before adding this to `mc-vr-api` proper.

In response to a lot of the ChatGPT stuff that makes no sense:
- 18 seconds is a ridiculous amount of time to store for, imo. Not only is that 360 positions (not 328 positions, as ChatGPT thinks for whatever reason), but I doubt anyone is going to need the velocity of a controller over 18 seconds. If they do for whatever reason, then honestly this might be a spot where the aforementioned method of each mod asking how many ticks it wants to go back would be useful. Maybe we could combine it with a guaranteed minimum that mods don't need to report (10 ticks I think would be safe here).
- 8KB for 328 positions implies that we use around 24.39 bytes per position (or around 24.98 bytes if it actually meant KiB), which makes zero sense. It's close, but in theory it should be giving something that exactly lines up on 24 bytes if it wanted to make sense.

Tl;dr: Average velocity wouldn't really work, and a lot of what was said was genuine junk. I _highly_ recommend not using ChatGPT for things, as it can confidently get answers very wrong.

**Comment by hammy275** [comment]
To add on a bit, if you want to simply get the average velocity for the past couple ticks, this is also relatively trivial to add to a given mod.

Here's the entire code that does the velocity calculations for ImmersiveMC while also storing data for the past 5 ticks (even though it really only needs to store the past 2) https://github.com/hammy3502/immersive-mc/blob/1.19.x/common/src/main/java/com/hammy275/immersivemc/client/LastClientVRData.java .

I would love to add an `IVRData` history at some point though. Would be nice so mods don't need to hold copies of the same data to calculate velocities off of. And of course, I do still very much want to provide a velocity API

**Comment by yohlimem** [comment]
The wrong data was all my fault sorry about that, ChatGPT just fixed my grammar and answered my question about velocity.
It looks like in the diagram you made, the velocity is just the two last positions since you are already moving in that direction.
the average velocity would be the red arrow because you're moving up and down so the vertical movement kinda cancels itself. Maybe for the average velocity you could find the position differences between each index and the one after it (e.g. i and i+1) and just divide them by the number of differences. So if you have four differences (0,2,3), (0,3,4), (3,4,2) and (0,1,2) you'd do  `(((0,3,4)-(0,2,3))+((3,4,2)-(0,3,4))+((0,1,2)-(3,4,2)))/4` .
[I made some nice code that does that]
(https://github.com/yohlimem/CalculateAverageVelocity/blob/main/src/main/java/net/fabricmc/av/features/FindPlayerVelocity.java)
In this example I used a cheesy way to calculate the velocity per second, I calculated the velocity per tick and multiplied by 20 to get the result I wanted. I think this works fine enough.
the thing is, for the player average velocity isn't very useful since the player can only apply acceleration in 9 directions and never moves unexpectedly but for a controller, I think this will work really well.

**Comment by hammy275** [comment]
My thought process would be that average velocity would be a good option, but wouldn't be the only option. The averaging option you proposed honestly isn't much better than regular average velocity, as if we had 4 points on the arc I posted earlier, we'd still end up with the same (arguably erroneous) result. I call this erroneous as actually trying to apply this velocity to something such as throwing would be wildly inaccurate.

I haven't looked into it, but the type of velocity I would be looking for would basically be doing a 3D line fit, then taking the derivative (instantaneous velocity if you want something to search off of) at the current tick. Would definitely be more mathematically heavy, but it could result in pretty accurate throwing, which would be nice. In the end, the API would look something like this:

`getAverageVelocity(int ticksAgo)`: Get the average velocity based on the position `ticksAgo` versus the position now.
`getPredictedVelocity(int ticksAgo)`: Gets the predicted velocity based on positions up to `ticksAgo` ticks ago.

The predicted velocity function might also take an `enum` for the type of prediction algorithm to use, as different applications or scenarios could lend themselves to different predictors (linear, logarithmic, etc.). At the same time, I may not have it, as simply making things easier for the developer may be a good idea, so it'll be worth trying to pick a best algorithm to use here. This is definitely an area where I'll need to do more research.

Tl;dr: Average velocity is good, but we should also add another option for more accuracy at the cost of computation.

**Comment by yohlimem** [comment]
I think these functions will be great additions to this API! Feel free to use my code if you'd like.
I think I will close this issue, it's getting pretty long.

**Comment by hammy275** [comment]
Going to re-open so I can track it until it's added to mc-vr-api proper.

**Comment by hammy275** [comment]
Merging this into #14, as the official API supports velocity math in-API.
